notifications/v1/client.go

```diff
@@ -0,0 +1,95 @@
+package v1
+
+import (
+	"context"
+	"encoding/json"
+
+	"cloud.google.com/go/pubsub"
+	"github.com/sourcegraph/log"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+// Client is a client for receiving SAMS notifications from a GCP Pub/Sub
+// subscription.
+type Client struct {
+	logger        log.Logger
+	subscription  *pubsub.Subscription
+	cancelContext context.CancelFunc
+}
+
+// NewClient creates a new Client for receiving SAMS notifications with given
+// GCP project ID and Pub/Sub subscription ID.
+func NewClient(logger log.Logger, projectID, subscriptionID string) (*Client, error) {
+	client, err := pubsub.NewClient(context.Background(), projectID)
+	if err != nil {
+		return nil, errors.Wrap(err, "create GCP Pub/Sub client")
+	}
+	return &Client{
+		logger:       logger.Scoped("notification.ClientV1"),
+		subscription: client.Subscription(subscriptionID),
+	}, nil
+}
+
+// ReceiveHandlers is a collection of receive handlers for each type of SAMS
+// notifications. If the handler of a notification is nil, the notification will
+// be acknowledged automatically without any processing.
+//
+// If a handler returns an error, the notification will be unacknowledged and
+// retried later.
+type ReceiveHandlers struct {
+	// OnUserDeleted is called when a "UserDeleted" notification is received.
+	OnUserDeleted func(data *UserDeletedData) error
+}
+
+// Receive starts receiving SAMS notifications and calls the corresponding
+// handler for each notification. It blocks until the context is done (e.g.
+// deadline exceed or canceled).
+//
+// ⚠️ WARNING: Each subscription can only have one active receiver at a time,
+// i.e. there should only be one call to Receive for a given subscription.
```

Comment:
I think this should panic if a handler has already been registered then

But, I also don't see a reason _not_ to support multiple receivers - we can have `[]handler` and iterate over those when receiving messages.

In that case you might want an API surface like:

```go
RegisterReceiver(handler)
Start() error
```

If you want to keep single-receiver-only, IMO you should add it to `NewClient`, then:

```
NewClient(...)
(c) Start() error
(c) Stop() error
```

Reply:
> So I'd (gently, humbly, kindly) suggest that we keep the SAMS SDK as simple as possible, to make it easier for clients to use safely/correctly. (And at least in this case, it doesn't seem like we are preventing more exotic/advanced use cases.)

Agree - I think adding `handler` as something you must provide up-front when creating a client is probably the best way to go, i.e. the API surface would be:

```go
NewClient(...) // provide handler here
(c) Start() error // long-lived routine subscribed to the events
(c) Stop() error // shutdown
```

Maybe `client` can just implement the background worker? Then it can start and stop with the MSP service runtime as part of the collection of routines :)

Reply:
> Maybe client can just implement the background job interface? Then it can start and stop with the MSP service runtime as part of the collection of routines :)

... that actually would be huge. 😍 

In the limit, every service eventually grows to the point that it needs to have some sort of "background job" / "message queue" system. And if there is some ultra-lightweight version that can be provided by the MSP runtime, that would be a capability that would be useful to MSP services in the future.

e.g. it is non-trivial to correctly/safely do the wiring to gracefully stop something. (e.g. when to call `Stop() error`, and make sure we call it in any number of situations -- e.g. SIGTERM, or some GCP-specific event, etc.) So being able to just delegate that to `runtime.RegisterShutdownCallback(logger, pubSubClient.Stop)` would mean that it's one less thing that we need to implement N-different times across various services.

Reply:
This already exists, you just have to return a multi-routine here https://sourcegraph.sourcegraph.com/github.com/sourcegraph/self-serve-cody/-/blob/cmd/accounts-server/service.go?L49-65, e.g. https://sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/cmd/telemetry-gateway/service/service.go?L119-138 - the runtime spins up the routine and does intercept handling and shutdown logic via `lib/background`, which runs all the single-tenant Sourcegraph services today

Reply:
Sorry for the confusion here 😅 I updated the docstring to explain that this constraint is imposed by `pubsub.Subscription.Receive` 😂 not me... as Chris said, if we were able to do it, it only makes the use case easier. For now, people can still create multiple `Client`s (that comes with different `pubsub.Subscription`s) to achieve the same effect if they _really_ want to.

---

On the topic of "Maybe client can just implement the background job interface", that's actually my plan, once https://github.com/sourcegraph/sourcegraph/pull/62136 is able to get merged (which may take another 2-3 weeks...)

Reply:
> But I do not want to force people to use background.Routine, so in my plan, the Client would return a ReceiveRoutine that implements the background.Routine they can grab and go.

Can we do this now, and just update this implementation later once https://github.com/sourcegraph/sourcegraph/pull/62136 lands? The current implementation IMO is hard to hold right - I think we should just have a new `ReceiverClient` that is a background routine, and users don't have to worry about starting and stopping it. Starting and stopping the receiver shouldn't be an application concern, it should match 1:1 with service lifecycle, is there ever a situation where it shouldn't?